File: llamafile/server/www/highlight_python.js
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const PYTHON_KEYWORDS = new Set([
  'and',
  'as',
  'assert',
  'async',
  'await',
  'break',
  'class',
  'continue',
  'def',
  'del',
  'elif',
  'else',
  'except',
  'finally',
  'for',
  'from',
  'global',
  'if',
  'import',
  'in',
  'is',
  'lambda',
  'nonlocal',
  'not',
  'or',
  'pass',
  'raise',
  'return',
  'try',
  'while',
  'with',
  'yield',
]);

const PYTHON_CONSTANTS = new Set([
  'False',
  'None',
  'True',
]);

const PYTHON_BUILTINS = new Set([
  '__import__',
  'abs',
  'aiter',
  'all',
  'anext',
  'any',
  'ascii',
  'bin',
  'bool',
  'breakpoint',
  'bytearray',
  'bytes',
  'callable',
  'chr',
  'classmethod',
  'compile',
  'complex',
  'delattr',
  'dict',
  'dir',
  'divmod',
  'enumerate',
  'eval',
  'exec',
  'filter',
  'float',
  'format',
  'frozenset',
  'getattr',
  'globals',
  'hasattr',
  'hash',
  'help',
  'hex',
  'id',
  'input',
  'int',
  'isinstance',
  'issubclass',
  'iter',
  'len',
  'list',
  'locals',
  'map',
  'max',
  'memoryview',
  'min',
  'next',
  'object',
  'oct',
  'open',
  'ord',
  'pow',
  'print',
  'property',
  'range',
  'repr',
  'reversed',
  'round',
  'set',
  'setattr',
  'slice',
  'sorted',
  'staticmethod',
  'str',
  'sum',
  'super',
  'tuple',
  'type',
  'vars',
  'zip',
]);

class HighlightPython extends Highlighter {

  static NORMAL = 0;
  static WORD = 1;
  static COM = 2;
  static SQUOTE = 3;
  static SQUOTESTR = 4;
  static SQUOTESTR_BACKSLASH = 5;
  static SQUOTE2 = 6;
  static SQUOTE3 = 7;
  static SQUOTE3_BACKSLASH = 8;
  static SQUOTE31 = 9;
  static SQUOTE32 = 10;
  static DQUOTE = 11;
  static DQUOTESTR = 12;
  static DQUOTESTR_BACKSLASH = 13;
  static DQUOTE2 = 14;
  static DQUOTE3 = 15;
  static DQUOTE3_BACKSLASH = 16;
  static DQUOTE31 = 17;
  static DQUOTE32 = 18;

  constructor(delegate) {
    super(delegate);
    this.word = '';
  }

  feed(input) {
    for (let i = 0; i < input.length; i += this.delta) {
      this.delta = 1;
      let c = input[i];
      switch (this.state) {

      case HighlightPython.NORMAL:
        if (!isascii(c) || isalpha(c) || c == '_') {
          this.epsilon(HighlightPython.WORD);
        } else if (c == '#') {
          this.state = HighlightPython.COM;
          this.push("span", "comment");
          this.append(c);
        } else if (c == '\'') {
          this.state = HighlightPython.SQUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == '"') {
          this.state = HighlightPython.DQUOTE;
          this.push("span", "string");
          this.append(c);
        } else {
          this.append(c);
        }
        break;

      case HighlightPython.WORD:
        if (!isascii(c) || isalnum(c) || c == '_') {
          this.word += c;
        } else {
          if (PYTHON_KEYWORDS.has(this.word)) {
            this.push("span", "keyword");
            this.append(this.word);
            this.pop();
          } else if (PYTHON_BUILTINS.has(this.word)) {
            this.push("span", "builtin");
            this.append(this.word);
            this.pop();
          } else if (PYTHON_CONSTANTS.has(this.word)) {
            this.push("span", "constant");
            this.append(this.word);
            this.pop();
          } else {
            this.append(this.word);
          }
          this.word = '';
          this.epsilon(HighlightPython.NORMAL);
        }
        break;

      case HighlightPython.COM:
        this.append(c);
        if (c == '\n') {
          this.pop();
          this.state = HighlightPython.NORMAL;
        }
        break;

      case HighlightPython.SQUOTE:
        this.append(c);
        if (c == '\'') {
          this.state = HighlightPython.SQUOTE2;
        } else if (c == '\\') {
          this.state = HighlightPython.SQUOTESTR_BACKSLASH;
        } else {
          this.state = HighlightPython.SQUOTESTR;
        }
        break;

      case HighlightPython.SQUOTESTR:
        this.append(c);
        if (c == '\'') {
          this.pop();
          this.state = HighlightPython.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightPython.SQUOTESTR_BACKSLASH;
        }
        break;

      case HighlightPython.SQUOTESTR_BACKSLASH:
        this.append(c);
        this.state = HighlightPython.SQUOTESTR;
        break;

      case HighlightPython.SQUOTE2:
        if (c == '\'') {
          this.append(c);
          this.state = HighlightPython.SQUOTE3;
        } else {
          this.pop();
          this.epsilon(HighlightPython.NORMAL);
        }
        break;

      case HighlightPython.SQUOTE3:
        this.append(c);
        if (c == '\'') {
          this.state = HighlightPython.SQUOTE31;
        } else if (c == '\\') {
          this.state = HighlightPython.SQUOTE3_BACKSLASH;
        }
        break;

      case HighlightPython.SQUOTE31:
        this.append(c);
        if (c == '\'') {
          this.state = HighlightPython.SQUOTE32;
        } else if (c == '\\') {
          this.state = HighlightPython.SQUOTE3_BACKSLASH;
        } else {
          this.state = HighlightPython.SQUOTE3;
        }
        break;

      case HighlightPython.SQUOTE32:
        this.append(c);
        if (c == '\'') {
          this.pop();
          this.state = HighlightPython.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightPython.SQUOTE3_BACKSLASH;
        } else {
          this.state = HighlightPython.SQUOTE3;
        }
        break;

      case HighlightPython.SQUOTE3_BACKSLASH:
        this.append(c);
        this.state = HighlightPython.SQUOTE3;
        break;

      case HighlightPython.DQUOTE:
        this.append(c);
        if (c == '"') {
          this.state = HighlightPython.DQUOTE2;
        } else if (c == '\\') {
          this.state = HighlightPython.DQUOTESTR_BACKSLASH;
        } else {
          this.state = HighlightPython.DQUOTESTR;
        }
        break;

      case HighlightPython.DQUOTESTR:
        this.append(c);
        if (c == '"') {
          this.pop();
          this.state = HighlightPython.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightPython.DQUOTESTR_BACKSLASH;
        }
        break;

      case HighlightPython.DQUOTESTR_BACKSLASH:
        this.append(c);
        this.state = HighlightPython.DQUOTESTR;
        break;

      case HighlightPython.DQUOTE2:
        if (c == '"') {
          this.append('"');
          this.state = HighlightPython.DQUOTE3;
        } else {
          this.pop();
          this.epsilon(HighlightPython.NORMAL);
        }
        break;

      case HighlightPython.DQUOTE3:
        this.append(c);
        if (c == '"') {
          this.state = HighlightPython.DQUOTE31;
        } else if (c == '\\') {
          this.state = HighlightPython.DQUOTE3_BACKSLASH;
        }
        break;

      case HighlightPython.DQUOTE31:
        this.append(c);
        if (c == '"') {
          this.state = HighlightPython.DQUOTE32;
        } else if (c == '\\') {
          this.state = HighlightPython.DQUOTE3_BACKSLASH;
        } else {
          this.state = HighlightPython.DQUOTE3;
        }
        break;

      case HighlightPython.DQUOTE32:
        this.append(c);
        if (c == '"') {
          this.pop();
          this.state = HighlightPython.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightPython.DQUOTE3_BACKSLASH;
        } else {
          this.state = HighlightPython.DQUOTE3;
        }
        break;

      case HighlightPython.DQUOTE3_BACKSLASH:
        this.append(c);
        this.state = HighlightPython.DQUOTE3;
        break;

      default:
        throw new Error('Invalid state');
      }
    }
  }

  flush() {
    switch (this.state) {
    case HighlightPython.WORD:
      if (PYTHON_KEYWORDS.has(this.word)) {
        this.push("span", "keyword");
        this.append(this.word);
        this.pop();
      } else if (PYTHON_BUILTINS.has(this.word)) {
        this.push("span", "builtin");
        this.append(this.word);
        this.pop();
      } else if (PYTHON_CONSTANTS.has(this.word)) {
        this.push("span", "constant");
        this.append(this.word);
        this.pop();
      } else {
        this.append(this.word);
      }
      this.word = '';
      break;
    case HighlightPython.COM:
    case HighlightPython.SQUOTE:
    case HighlightPython.SQUOTESTR:
    case HighlightPython.SQUOTESTR_BACKSLASH:
    case HighlightPython.SQUOTE2:
    case HighlightPython.SQUOTE3:
    case HighlightPython.SQUOTE3_BACKSLASH:
    case HighlightPython.SQUOTE31:
    case HighlightPython.SQUOTE32:
    case HighlightPython.DQUOTE:
    case HighlightPython.DQUOTESTR:
    case HighlightPython.DQUOTESTR_BACKSLASH:
    case HighlightPython.DQUOTE2:
    case HighlightPython.DQUOTE3:
    case HighlightPython.DQUOTE3_BACKSLASH:
    case HighlightPython.DQUOTE31:
    case HighlightPython.DQUOTE32:
      this.pop();
      break;
    default:
      break;
    }
    this.state = HighlightPython.NORMAL;
    this.delegate.flush();
    this.delta = 1;
  }
}

Highlighter.REGISTRY['python'] = HighlightPython;
Highlighter.REGISTRY['py'] = HighlightPython;

File: llamafile/highlight/highlight_python.cpp
// -*- mode:c++;indent-tabs-mode:nil;c-basic-offset:4;coding:utf-8 -*-
// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
//
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "highlight.h"

#include <ctype.h>

enum
{
    NORMAL,
    WORD,
    COM,
    SQUOTE,
    SQUOTESTR,
    SQUOTESTR_BACKSLASH,
    SQUOTE2,
    SQUOTE3,
    SQUOTE3_BACKSLASH,
    SQUOTE31,
    SQUOTE32,
    DQUOTE,
    DQUOTESTR,
    DQUOTESTR_BACKSLASH,
    DQUOTE2,
    DQUOTE3,
    DQUOTE3_BACKSLASH,
    DQUOTE31,
    DQUOTE32,
};

HighlightPython::HighlightPython()
{
}

HighlightPython::~HighlightPython()
{
}

void
HighlightPython::feed(std::string* r, std::string_view input)
{
    int c;
    for (size_t i = 0; i < input.size(); ++i) {
        c = input[i] & 255;
        switch (t_) {

        Normal:
        case NORMAL:
            if (!isascii(c) || isalpha(c) || c == '_') {
                t_ = WORD;
                goto Word;
            } else if (c == '#') {
                t_ = COM;
                *r += HI_COMMENT;
                *r += c;
            } else if (c == '\'') {
                t_ = SQUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == '"') {
                t_ = DQUOTE;
                *r += HI_STRING;
                *r += c;
            } else {
                *r += c;
            }
            break;

        Word:
        case WORD:
            if (!isascii(c) || isalnum(c) || c == '_') {
                word_ += c;
            } else {
                if (is_keyword_python(word_.data(), word_.size())) {
                    *r += HI_KEYWORD;
                    *r += word_;
                    *r += HI_RESET;
                } else if (is_keyword_python_builtin(word_.data(),
                                                     word_.size())) {
                    *r += HI_BUILTIN;
                    *r += word_;
                    *r += HI_RESET;
                } else if (is_keyword_python_constant(word_.data(),
                                                      word_.size())) {
                    *r += HI_CONSTANT;
                    *r += word_;
                    *r += HI_RESET;
                } else {
                    *r += word_;
                }
                word_.clear();
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case COM:
            *r += c;
            if (c == '\n') {
                *r += HI_RESET;
                t_ = NORMAL;
            }
            break;

        // handle 'string'
        case SQUOTE:
            *r += c;
            if (c == '\'') {
                t_ = SQUOTE2;
            } else if (c == '\\') {
                t_ = SQUOTESTR_BACKSLASH;
            } else {
                t_ = SQUOTESTR;
            }
            break;
        case SQUOTESTR:
            *r += c;
            if (c == '\'') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = SQUOTESTR_BACKSLASH;
            }
            break;
        case SQUOTESTR_BACKSLASH:
            *r += c;
            t_ = SQUOTESTR;
            break;

        // handle '''string'''
        case SQUOTE2:
            if (c == '\'') {
                *r += c;
                t_ = SQUOTE3;
            } else {
                *r += HI_RESET;
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case SQUOTE3:
            *r += c;
            if (c == '\'') {
                t_ = SQUOTE31;
            } else if (c == '\\') {
                t_ = SQUOTE3_BACKSLASH;
            }
            break;

        case SQUOTE31:
            *r += c;
            if (c == '\'') {
                t_ = SQUOTE32;
            } else if (c == '\\') {
                t_ = SQUOTE3_BACKSLASH;
            } else {
                t_ = SQUOTE3;
            }
            break;

        case SQUOTE32:
            *r += c;
            if (c == '\'') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = SQUOTE3_BACKSLASH;
            } else {
                t_ = SQUOTE3;
            }
            break;

        case SQUOTE3_BACKSLASH:
            *r += c;
            t_ = SQUOTE3;
            break;

        // handle "string"
        case DQUOTE:
            *r += c;
            if (c == '"') {
                t_ = DQUOTE2;
            } else if (c == '\\') {
                t_ = DQUOTESTR_BACKSLASH;
            } else {
                t_ = DQUOTESTR;
            }
            break;

        case DQUOTESTR:
            *r += c;
            if (c == '"') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = DQUOTESTR_BACKSLASH;
            }
            break;

        case DQUOTESTR_BACKSLASH:
            *r += c;
            t_ = DQUOTESTR;
            break;

        // handle """string"""
        case DQUOTE2:
            if (c == '"') {
                *r += '"';
                t_ = DQUOTE3;
            } else {
                *r += HI_RESET;
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case DQUOTE3:
            *r += c;
            if (c == '"') {
                t_ = DQUOTE31;
            } else if (c == '\\') {
                t_ = DQUOTE3_BACKSLASH;
            }
            break;

        case DQUOTE31:
            *r += c;
            if (c == '"') {
                t_ = DQUOTE32;
            } else if (c == '\\') {
                t_ = DQUOTE3_BACKSLASH;
            } else {
                t_ = DQUOTE3;
            }
            break;

        case DQUOTE32:
            *r += c;
            if (c == '"') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = DQUOTE3_BACKSLASH;
            } else {
                t_ = DQUOTE3;
            }
            break;

        case DQUOTE3_BACKSLASH:
            *r += c;
            t_ = DQUOTE3;
            break;

        default:
            __builtin_unreachable();
        }
    }
}

void
HighlightPython::flush(std::string* r)
{
    switch (t_) {
    case WORD:
        if (is_keyword_python(word_.data(), word_.size())) {
            *r += HI_KEYWORD;
            *r += word_;
            *r += HI_RESET;
        } else if (is_keyword_python_builtin(word_.data(), word_.size())) {
            *r += HI_BUILTIN;
            *r += word_;
            *r += HI_RESET;
        } else if (is_keyword_python_constant(word_.data(), word_.size())) {
            *r += HI_CONSTANT;
            *r += word_;
            *r += HI_RESET;
        } else {
            *r += word_;
        }
        word_.clear();
        break;
    case COM:
    case SQUOTE:
    case SQUOTESTR:
    case SQUOTESTR_BACKSLASH:
    case SQUOTE2:
    case SQUOTE3:
    case SQUOTE3_BACKSLASH:
    case SQUOTE31:
    case SQUOTE32:
    case DQUOTE:
    case DQUOTESTR:
    case DQUOTESTR_BACKSLASH:
    case DQUOTE2:
    case DQUOTE3:
    case DQUOTE3_BACKSLASH:
    case DQUOTE31:
    case DQUOTE32:
        *r += HI_RESET;
        break;
    default:
        break;
    }
    t_ = NORMAL;
}
